File: packages/core/src/simulation/scenario.ts
import { VenueGraph } from '../routing/graph.js';
import { Simulation, type Leg, type SimConfig } from './model.js';

export interface Cohort {
  count: number;
  origin: string;
  destination: string;
  start_s: number;
  spread_s: number;
  itinerary?: Leg[];
}
export class Scenario {
  populate: ((simulation: Simulation) => void) | null = null;

  constructor(
    readonly name: string,
    readonly description: string,
    readonly cohorts: Cohort[],
    readonly durationS = 900,
    readonly seed = 42,
  ) {}
  build(graph: VenueGraph, overrides: Partial<SimConfig> = {}): Simulation {
    const simulation = new Simulation(graph, { seed: this.seed, ...overrides });
    if (this.populate) {
      this.populate(simulation);
      return simulation;
    }
    for (const cohort of this.cohorts) {
      if (cohort.itinerary?.length)
        simulation.addItinerary(cohort.count, cohort.origin, cohort.itinerary, cohort.start_s, cohort.spread_s);
      else simulation.addAgents(cohort.count, cohort.origin, cohort.destination, cohort.start_s, cohort.spread_s);
    }
    return simulation;
  }
}

export function arrival(
  _graph: VenueGraph,
  gate: string,
  stand: string,
  count = 1500,
  seed = 42,
  spreadS = 300,
): Scenario {
  return new Scenario(
    'arrival',
    `${count} spectators arrive through ${gate} for ${stand}`,
    [{ count, origin: gate, destination: stand, start_s: 0, spread_s: spreadS }],
    1800,
    seed,
  );
}

export function egress(
  _graph: VenueGraph,
  origins: string[] | string,
  exit: string,
  count = 2000,
  seed = 42,
  spreadS = 240,
): Scenario {
  const list = typeof origins === 'string' ? [origins] : origins;
  if (!list.length) throw new Error('egress needs at least one origin');
  const base = Math.trunc(count / list.length);
  const remainder = count % list.length;
  const cohorts = list
    .map((origin, index) => ({
      count: base + (index < remainder ? 1 : 0),
      origin,
      destination: exit,
      start_s: 0,
      spread_s: spreadS,
    }))
    .filter((cohort) => cohort.count > 0);
  return new Scenario(
    'post-race-egress',
    `${count} spectators leave ${cohorts.length} stand(s) for ${exit} at the flag`,
    cohorts,
    1800,
    seed,
  );
}

File: packages/core/src/crowdflow_core/simulation/scenario.py
"""Scenarios.

A scenario is a seeded, declarative description of demand — not a script that
forces a bottleneck. Congestion has to emerge from demand meeting capacity, or
the prediction it validates is circular.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..routing.graph import VenueGraph
from .model import SimConfig, Simulation


@dataclass(frozen=True)
class Cohort:
    count: int
    origin: str
    destination: str
    start_s: float = 0.0
    spread_s: float = 0.0


@dataclass(frozen=True)
class Scenario:
    name: str
    description: str
    cohorts: list[Cohort] = field(default_factory=list)
    duration_s: float = 900.0
    seed: int = 42

    def build(self, graph: VenueGraph, **overrides) -> Simulation:
        config = SimConfig(seed=self.seed, **overrides)
        sim = Simulation(graph, config)
        for c in self.cohorts:
            sim.add_agents(
                c.count, c.origin, c.destination,
                start_s=c.start_s, spread_s=c.spread_s,
            )
        return sim


def egress(graph: VenueGraph, origins: list[str] | str, exit_zone: str,
           count: int = 2000, seed: int = 42, spread_s: float = 240.0) -> Scenario:
    """The hardest twenty minutes of the weekend: everyone leaves at once.

    Not a gentle ramp — a step. Post-race egress is the largest crowd movement of
    the day and the one an operator most needs warning of.

    Takes several origins on purpose. A real crowd leaves from every grandstand
    simultaneously; funnelling the whole population out of a single node produces
    a queue at that node and tells you nothing about the venue, which is a
    property of the scenario rather than of the venue.
    """
    if isinstance(origins, str):
        origins = [origins]
    per = max(1, count // len(origins))
    return Scenario(
        name="post-race-egress",
        description=(
            f"{per * len(origins)} spectators leave {len(origins)} stand(s) "
            f"for {exit_zone} at the flag"
        ),
        cohorts=[
            Cohort(count=per, origin=o, destination=exit_zone,
                   start_s=0.0, spread_s=spread_s)
            for o in origins
        ],
        duration_s=1800.0,
        seed=seed,
    )


def arrival(graph: VenueGraph, gate: str, stand: str, count: int = 1500,
            seed: int = 42) -> Scenario:
    """Gates open. Spread over an hour, so this should stay nominal — a scenario
    that never congests is as useful a test as one that does."""
    return Scenario(
        name="arrival",
        description=f"{count} spectators arrive at {gate} for {stand}",
        cohorts=[Cohort(count=count, origin=gate, destination=stand,
                        start_s=0.0, spread_s=3600.0)],
        duration_s=1800.0,
        seed=seed,
    )
